### ISSUE TITLE
java.io.FileNotFoundException: .clj-kondo/.cache/$version/lock

### ISSUE THREAD

**Issue by andrezacm** [borkdude/clj-kondo#542]
**version**
v2019.10.11-alpha

**platform**
macOS v10.14.6

**problem**
Unexpected error when a project has a .clj-kondo directory

**repro**
`brew install borkdude/brew/clj-kondo`
`mkdir .clj-kondo` (in the root of my project)
`clj-kondo --lint "$(lein classpath)"`

**expected behavior**
Same report I have when running the same thing without .clj-kondo directory

**actual behavior**
```
clj-kondo --lint "$(lein classpath)"                                                                                                     add-new-queues-for-instability!
:repositories detected in user-level profiles! [:user] 
See https://github.com/technomancy/leiningen/wiki/Repeatability
Unexpected error. Please report an issue.
java.io.FileNotFoundException: /Users/<path-of-my-project>/.clj-kondo/.cache/2019.10.11-alpha/lock
	at com.oracle.svm.core.posix.PosixUtils.fileOpen(PosixUtils.java:235)
	at java.io.RandomAccessFile.open(RandomAccessFile.java:709)
	at java.io.RandomAccessFile.<init>(RandomAccessFile.java:243)
	at clj_kondo.impl.cache$sync_cache.invokeStatic(cache.clj:123)
	at clj_kondo.core$run_BANG_.invokeStatic(core.clj:97)
	at clj_kondo.main$main.invokeStatic(main.clj:85)
	at clj_kondo.main$main.doInvoke(main.clj:78)
	at clojure.lang.RestFn.applyTo(RestFn.java:137)
	at clojure.core$apply.invokeStatic(core.clj:665)
	at clj_kondo.main$_main$fn__4930.invoke(main.clj:104)
	at clj_kondo.main$_main.invokeStatic(main.clj:104)
	at clj_kondo.main$_main.doInvoke(main.clj:102)
	at clojure.lang.RestFn.applyTo(RestFn.java:137)
	at clj_kondo.main.main(Unknown Source)
```

**Comment by borkdude** [comment]
@andrezacm Thanks for reporting! I have one other user report this on Slack, but I wasn't able to reproduce it then. 

- Can you make a minimal copy of your project where the error still occurs, with all sensitive information removed and attach it as a .zip here? I hope I can reproduce on my own system.

- Can you try linting in the same way as above with the uberjar? That is available here:
https://github.com/borkdude/clj-kondo/releases/download/v2019.10.11-alpha/clj-kondo-2019.10.11-alpha-standalone.jar

**Comment by borkdude** [comment]
@andrezacm I've tried to repro this on my own system where I'm also using MacOS:

```
$ lein new lock
$ cd lock
$ mkdir .clj-kondo
$ clj-kondo --lint $(lein classpath)
... program exits normally
$ ls -la .clj-kondo/.cache/2019.10.12-alpha-SNAPSHOT/lock
-rw-r--r--  1 borkdude  wheel  0 Oct 22 13:16 .clj-kondo/.cache/2019.10.12-alpha-SNAPSHOT/lock
```

Can you maybe also try these steps?

**Comment by andrezacm** [comment]
Thanks for answering @borkdude, I've reproduced the steps with a brand new project in the same computer and it worked perfectly, although the project related it's from the company I work for I can't send it here, I'm going to try in other projects though. We could talk on slack too if it's easier.

**Comment by borkdude** [comment]
From Slack:

It turns out that linting the project sources go well, but the problem is with one of the dependencies from the classpath. Yet to be determined which dependencies are problematic.

**Comment by andrezacm** [comment]
From slack: 
After removing corporate dependencies from classpath it is running well.

**Comment by markdingram** [comment]
Seems to be related to the size of the classpath - copy pasted a few projects dependencies together to give a reproducible project - https://gist.github.com/markdingram/b6898f2b830c76774deb1fb09bd28548

As it stands this project consistently fails. 

From a quick look doesn't seem to be 1 dependency in particular as the root cause - removing different deps leads to a successful run.

**Comment by borkdude** [comment]
@markdingram Thanks for that, I can repro. I will include that project.clj in a comment here:

[project.clj.zip](https://github.com/borkdude/clj-kondo/files/3799357/project.clj.zip)

So using `clj-kondo --lint $(lein classpath)` I get the above error, but not using the standalone jar:

```
$ java -jar clj-kondo-2019.10.26-standalone.jar --lint "$(lein classpath)"  > /tmp/log.txt
```

Maybe it's a GraalVM related issue. I will search in that direction.

The relevant code is here: https://github.com/borkdude/clj-kondo/blob/9de6b455a8ccd6df7923cb480b621c42572908a1/src/clj_kondo/impl/cache.clj#L65
A `RandomAccessFile` is created using `rw`. This should only except with a `FileNotFoundException` under this condition according to its documentation:

> if the mode begins with "rw" but the given string does not denote an existing, writable regular file and a new regular file of that name cannot be created, or if some other error occurs while opening or creating the file

@andrezacm Could you test if the standalone jar works for linting your classpath? You can download it on the releases page: 

https://github.com/borkdude/clj-kondo/releases/download/v2019.10.26/clj-kondo-2019.10.26-standalone.jar

**Comment by borkdude** [comment]
I created a GraalVM issue [here](https://github.com/oracle/graal/issues/1804).

**Comment by borkdude** [comment]
@markdingram Are you also using MacOS? Which version of MacOS?

**Comment by markdingram** [comment]
Yes, also 10.14.6

**Comment by borkdude** [comment]
Interesting, so far I've only had macOS people mention this problem and one linux user who could not reproduce it.

**Comment by borkdude** [comment]
@andrezacm @markdingram Found the cause and hopefully fixed! The problem was then when opening JarFiles, they were never closed and apparently weird things happen in the binary version when you have too many of them open.

This new Mac binary should fix the problem:

https://6819-176829714-gh.circle-artifacts.com/0/release/clj-kondo-2019.10.27-SNAPSHOT-macos-amd64.zip
